Add Key Vault service into fog-azure-rm

vishrutshah opened the pull request:
As part of understanding the `fog-azure-rm` source code and model, I was building a new Key Vault service into fog-azure-rm. 

It's be great to get you guys feedback & suggestions on this PR. We'd appreciate it.

Thanks,

bilal-naeem-confiz commented on lib/fog/azurerm/models/key_vault/vault.rb:
```diff
@@ -0,0 +1,81 @@
+        attribute :enabled_for_template_deployment
+
+        def self.parse(vault)
+          puts vault
```
Remove this puts statement

bilal-naeem-confiz commented on lib/fog/azurerm/models/key_vault/vault.rb:
```diff
@@ -0,0 +1,81 @@
+            end
+
+            vault_hash['access_policies'] = []
+            vault_properties.access_policies.each do |access_policy|
```
Check if access policies are nil and then proceed to access the values

bilal-naeem-confiz commented on lib/fog/azurerm/models/key_vault/vault.rb:
```diff
@@ -0,0 +1,81 @@
+
+        private
+
+        def validate_input(attr_hash)
```
If this method is not being used and has no body then it is best to remove it

bilal-naeem-confiz commented on lib/fog/azurerm/requests/key_vault/create_or_update_vault.rb:
```diff
@@ -0,0 +1,103 @@
+            }
+          }
+          vault_mapper = Azure::ARM::KeyVault::Models::Vault.mapper
+          @kay_vault_client.deserialize(vault_mapper, vault, 'result.body')
```
Typo in ``` @kay_vault_client ```
Should be ``` @key_vault_client ```

bilal-naeem-confiz commented on lib/fog/azurerm/requests/key_vault/get_vault.rb:
```diff
@@ -0,0 +1,54 @@
+            }
+          }
+          vault_mapper = Azure::ARM::KeyVault::Models::Vault.mapper
+          @kay_vault_client.deserialize(vault_mapper, vault, 'result.body')
```
Typo in ``` @kay_vault_client ```
Should be ``` @key_vault_client ```

bilal-naeem-confiz commented on lib/fog/azurerm/requests/key_vault/list_vaults.rb:
```diff
@@ -0,0 +1,56 @@
+            }
+          ]
+          vaults_mapper = Azure::ARM::KeyVault::Models::VaultListResult.mapper
+          @kay_vault_client.deserialize(vaults_mapper, vaults, 'result.body')
```
Typo in ``` @kay_vault_client ```
Should be ``` @key_vault_client ```

bilal-naeem-confiz commented:
Hi @vishrutshah 

These are the changes in the code-base. We suggest you to fix these in the relevant places

1. There are no unit tests. Please write unit tests for the code and ensure maximum code coverage.

2. The complete namespace of a model should be written instead of the model name. This ensures that there will be no conflicts with existing providers written in fog e.g. instead of ``` Vault ``` write ``` Fog::KeyVault:::AzureRM:::Vault ``` in [this](https://github.com/fog/fog-azure-rm/pull/217/files#diff-a4186909b6a31db99602bf50a7e07148) file

3. Integration tests contain all the functionality that is provided to the user. Make sure to write use cases of each functionality in the integration tests. This includes writing use case for ``` get ```, ``` list ``` and ``` delete ``` requests for KeyVault.

4. Add proper documentation for the feature that you have added to fog. This documentation should be as a markdown file and should be added to ``` lib/fog/azurerm/docs ```

5. In most models implemented in fog, we use ` name ` as the identity of the model. Please use ` name ` as the identity in [models/key_vault/access_policy_entry.rb](https://github.com/fog/fog-azure-rm/pull/217/files#diff-51460fa55f5fb6f0f461bb502ab6acc8)

Other issues are specified as comments on the code. We will review your changes as soon as you are done.

Thanks

vishrutshah commented:
@bilal-naeem-confiz Awesome .Thanks you very much for taking time to go through the first pass. I'll work on addressing the comments and update this PR soon. Thanks a lot again!

vishrutshah commented:
Hello @bilal-naeem-confiz,

I am still working on writing more unit tests and increasing the coverage & I've addressed some of the issues you mentioned in review and points. I have couple of questions on the 5 points that you mentioned:

Regarding point 5, [AccessPolicyEntry](https://github.com/Azure/azure-sdk-for-ruby/blob/master/management/azure_mgmt_key_vault/lib/generated/azure_mgmt_key_vault/models/access_policy_entry.rb) model of `Azure::ARM::KeyVault::Models` dows not have any attribute named as `name`, could you please advise how can i then use `name` as the identity of the model if there is a way or what should we prefer in that case?

Regarding point 2, could you please elaborate as I am not clear on how should i use full namespace & in which file? 

Thanks!

bilal-naeem-confiz commented:
Hi @vishrutshah 

Regarding point 2, what I mean is that instead of just using the model name like done [here](https://github.com/fog/fog-azure-rm/pull/217/files#diff-a4186909b6a31db99602bf50a7e07148R35) you should instead use it like ``` Fog::KeyVault::AzureRM::Vault.new ``` The reason for using this is to avoid conflicts with other cloud providers integrated with fog.
Regarding point 5, even if the Azure Ruby SDK does not show ``` name ``` as an attribute, the attribute exists for all resources. Attributes like ``` name ```, ``` resource_group ``` and ``` location ``` are all part of ``` MsRestAzure::Resource ``` and are inherited by all models present in the SDK.

I hope this answers any queries you had. Feel free to ask about any other issues that arise.

Thanks

vishrutshah commented:
@bilal-naeem-confiz Thank you so much for the quick response. 

I've added unit tests for models & requests, added usage documentations and updated integration tests. I've also fixed the comments and feedbacks you have provided on the review. 

Now, that only think is I am still having trouble understating point 5. :) I understand that you are mentioning that attributes like `name`, `resource_group` and `location` are all part of [MsRestAzure::Resource ](https://github.com/Azure/azure-sdk-for-ruby/blob/master/runtime/ms_rest_azure/lib/ms_rest_azure/resource.rb#L9). The way I am understanding is each model's properties are supposed to go on-wire to the server but that doesn't seem to be the case here because we are not able create [AccessPolicyEntry ](https://github.com/Azure/azure-sdk-for-ruby/blob/master/management/azure_mgmt_key_vault/lib/generated/azure_mgmt_key_vault/models/access_policy_entry.rb#L12) with name. So if we are not going to send it on the wire & not required by the server for AccessPolicyEntry, why should we require `name` from the user / having it as identity to build upon? May be I am not getting the point of having identity in fog but it'd be great if you can please help me with examples to understand. :)

Also, Feel free to review & provide more feedbacks as you guys get a chance. Thanks!

bilal-naeem-confiz commented:
Hi @vishrutshah 

I see that you have only written unit test for ``` create_or_update ``` request. Please write unit tests for other requests such as ``` get ```, ``` list ``` and ``` delete ``` to ensure maximum code coverage.

Thanks

vishrutshah commented:
@bilal-naeem-confizthanks for catching that. I totally missed that. Added those tests. Thanks!

vishrutshah commented:
Hi @bilal-naeem-confiz, I've rebased my changes on top of master due to conflicts. Please review and let me know your views on this. 

Thanks,
Vishrut Shah
